src/crypto/qat/qcccrypto.cc

```diff
@@ -20,51 +26,111 @@ static std::ostream& _prefix(std::ostream* _dout)
 // -----------------------------------------------------------------------------
 
 /*
- * Poller thread & functions
-*/
+ * Callback function
+ */
+static void symDpCallback(CpaCySymDpOpData *pOpData,
+                        CpaStatus status,
+                        CpaBoolean verifyResult)
+{
+  if (nullptr != pOpData->pCallbackTag)
+  {
+    if (static_cast<QatCrypto*>(pOpData->pCallbackTag)->complete()) {
+      static_cast<QatCrypto*>(pOpData->pCallbackTag)->completion_handler(std::error_code{});
+    }
+  }
+}
+
 static std::mutex qcc_alloc_mutex;
 static std::mutex qcc_eng_mutex;
 static std::atomic<bool> init_called = { false };
 
-void* QccCrypto::crypt_thread(void *args) {
-  struct qcc_thread_args *thread_args = (struct qcc_thread_args *)args;
-  thread_args->qccinstance->do_crypt(thread_args);
-  return thread_args;
+#define NON_INSTANCE -1
+#define RETRY_MAX_NUM 100
+
+template <typename CompletionToken>
+auto QccCrypto::async_get_instance(CompletionToken&& token) {
+  using boost::asio::async_completion;
+  using Signature = void(int);
+  async_completion<CompletionToken, Signature> init(token);
+
+  auto ex = boost::asio::get_associated_executor(init.completion_handler);
+  static size_t max_queue_size = max_requests - qcc_inst->num_instances;
```

Comment:
this shouldn't be static. it could be a member variable instead, initialized by `init()`

also needs to protect against underflow in case `max_requests < qcc_inst->num_instances`

Reply:
Done. I use a member variable `max_queue_size` in `QccCrypto`, and initialized by `init()`.  And here doesn't need to protect against underflow, because `max_queue_size` means the wait request queue, not all request queue.